components/modal/frame.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * External dependencies
+ */
+import clickOutside from 'react-click-outside';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, compose, createRef } from '@wordpress/element';
+import { focus, keycodes } from '@wordpress/utils';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import withFocusReturn from '../higher-order/with-focus-return';
+import withFocusContain from '../higher-order/with-focus-contain';
+import withGlobalEvents from '../higher-order/with-global-events';
+import withSafeTimeout from '../higher-order/with-safe-timeout';
+
+const {
+	ESCAPE,
+} = keycodes;
+
+class ModalFrame extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.containerRef = createRef();
+		this.handleKeyDown = this.handleKeyDown.bind( this );
+		this.handleClickOutside = this.handleClickOutside.bind( this );
+		this.focusFirstTabbable = this.focusFirstTabbable.bind( this );
+	}
+
+	componentDidMount() {
+		// Focus on mount
+		if ( this.props.focusOnMount ) {
+			this.focusFirstTabbable();
+		}
+	}
+
+	focusFirstTabbable() {
+		// Required because the node is appended to the DOM after rendering.
+		const { setTimeout } = this.props;
+		setTimeout( () => {
+			const tabbables = focus.tabbable.find( this.containerRef.current );
+			if ( tabbables.length ) {
+				tabbables[ 0 ].focus();
+			}
+		}, 0 );
+	}
+
+	handleClickOutside( event ) {
+		if ( this.props.shouldCloseOnClickOutside ) {
+			this.onRequestClose( event );
+		}
+	}
+
+	handleKeyDown( event ) {
+		if ( event.keyCode === ESCAPE ) {
+			this.handleEscapeKeyDown( event );
+		}
+	}
+
+	handleEscapeKeyDown( event ) {
+		if ( this.props.shouldCloseOnEsc ) {
+			event.preventDefault();
+			this.onRequestClose( event );
+		}
+	}
+
+	onRequestClose( event ) {
+		const { onRequestClose } = this.props;
+		if ( onRequestClose ) {
+			onRequestClose( event );
+		}
+	}
+
+	render() {
+		const {
+			contentLabel,
+			aria: {
+				describedby,
+				labelledby,
+			},
+			children,
+			className,
+			style,
+		} = this.props;
+
+		return (
+			<div
+				className={ className }
+				style={ style }
+				ref={ this.containerRef }
+				role="dialog"
+				aria-modal={ true }
```

Comment:
For some reason, `aria-modal="true"` prevents Safari + VoiceOver to read out the content of the modal when using Control-Option-Arrows. This doesn't happen on `react-modal` because it (incorrectly` sets it on the overlay.  `aria-modal` is new in ARIA 1.1 and apparently not well supported. We should investigate this, since its effect is supposed to be the same we get with `aria-hidden` on the content outside the modal, maybe we should just not use it for now.